dbms/tests/queries/0_stateless/00597_push_down_predicate.sql

```diff
@@ -1,82 +1,110 @@
 SELECT 1;
 SELECT 1 AS id WHERE id = 1;
 SELECT arrayJoin([1,2,3]) AS id WHERE id = 1;
-SELECT * FROM (SELECT perf_1.z AS z_1 FROM test.perf AS perf_1);
+SELECT * FROM test.test WHERE id = 1;
 
 SELECT '-------Need push down-------';
+
+-- Optimize predicate expressions without tables
+ANALYZE SELECT * FROM system.one ANY LEFT JOIN (SELECT 0 AS dummy) USING dummy WHERE 1;
 SELECT * FROM system.one ANY LEFT JOIN (SELECT 0 AS dummy) USING dummy WHERE 1;
+
+ANALYZE SELECT toString(value) AS value FROM (SELECT 1 AS value) WHERE value = '1';
 SELECT toString(value) AS value FROM (SELECT 1 AS value) WHERE value = '1';
+
+ANALYZE SELECT * FROM (SELECT 1 AS id UNION ALL SELECT 2) WHERE id = 1;
 SELECT * FROM (SELECT 1 AS id UNION ALL SELECT 2) WHERE id = 1;
+
+ANALYZE SELECT * FROM (SELECT arrayJoin([1, 2, 3]) AS id) WHERE id = 1;
 SELECT * FROM (SELECT arrayJoin([1, 2, 3]) AS id) WHERE id = 1;
+
+ANALYZE SELECT id FROM (SELECT arrayJoin([1, 2, 3]) AS id) WHERE id = 1;
 SELECT id FROM (SELECT arrayJoin([1, 2, 3]) AS id) WHERE id = 1;
-SELECT * FROM (SELECT perf_1.z AS z_1 FROM test.perf AS perf_1) WHERE z_1 = 1;
 
+ANALYZE SELECT * FROM (SELECT 1 AS id, (SELECT 1) as subquery) WHERE subquery = 1;
 SELECT * FROM (SELECT 1 AS id, (SELECT 1) as subquery) WHERE subquery = 1;
+
+-- Optimize predicate expressions using tables
+ANALYZE SELECT * FROM (SELECT toUInt64(b) AS a, sum(id) AS b FROM test.test) WHERE a = 3;
 SELECT * FROM (SELECT toUInt64(b) AS a, sum(id) AS b FROM test.test) WHERE a = 3;
-SELECT * FROM (SELECT toUInt64(b), sum(id) AS b FROM test.test) WHERE `toUInt64(sum(id))` = 3;
+
+ANALYZE SELECT date, id, name, value FROM (SELECT date, name, value, min(id) AS id FROM test.test GROUP BY date, name, value) WHERE id = 1;
 SELECT date, id, name, value FROM (SELECT date, name, value, min(id) AS id FROM test.test GROUP BY date, name, value) WHERE id = 1;
-SELECT * FROM (SELECT toUInt64(b) AS a, sum(id) AS b FROM test.test AS table_alias) AS outer_table_alias WHERE outer_table_alias.b = 3;
 
-SELECT '-------Force push down-------';
-SET force_primary_key = 1;
+ANALYZE SELECT * FROM (SELECT toUInt64(b) AS a, sum(id) AS b FROM test.test AS table_alias) AS outer_table_alias WHERE outer_table_alias.b = 3;
+SELECT * FROM (SELECT toUInt64(b) AS a, sum(id) AS b FROM test.test AS table_alias) AS outer_table_alias WHERE outer_table_alias.b = 3;
 
 -- Optimize predicate expression with asterisk
+ANALYZE SELECT * FROM (SELECT * FROM test.test) WHERE id = 1;
 SELECT * FROM (SELECT * FROM test.test) WHERE id = 1;
+
 -- Optimize predicate expression with asterisk and nested subquery
+-- FIXME: should be pushed down to the innermost subquery
+ANALYZE SELECT * FROM (SELECT * FROM (SELECT * FROM test.test)) WHERE id = 1;
 SELECT * FROM (SELECT * FROM (SELECT * FROM test.test)) WHERE id = 1;
+
 -- Optimize predicate expression with qualified asterisk
+-- FIXME: should be pushed down to the innermost subquery
```

Comment:
This will be pushed down again when interpreting the subquery. maybe need to recurse [here](https://github.com/yandex/ClickHouse/pull/4387/files#diff-66265ce0e53c290ab1e8377eaaa9c0d4R88) ?